lib/CL/devices/pthread/pthread.c

```diff
@@ -117,6 +113,10 @@ struct data {
   mem_regions_management* mem_regions;
 #endif
 
+  /* Lock for command list related operations */
+  pthread_mutex_t cq_lock __attribute__ ((aligned (CACHELINE_SIZE)));
```

Comment:
OK. Surprising indeed. Volatiles should not cause accesses to main memory / cache flushs, only avoid storing it to registers, AFAIK. I think one still can cache them, cache coherence HW takes care of the consistency.